crates/partition-store/src/persisted_lsn_tracking.rs

```diff
@@ -0,0 +1,263 @@
+use std::ffi::{CStr, CString};
+use std::sync::Arc;
+
+use ahash::HashMap;
+use dashmap::DashMap;
+use rocksdb::event_listener::{EventListener, FlushJobInfo};
+use rocksdb::table_properties::{
+    CollectorError, EntryType, TablePropertiesCollector, TablePropertiesCollectorFactory,
+};
+use serde::{Deserialize, Serialize};
+use tracing::warn;
+
+use restate_storage_api::StorageError;
+use restate_types::{identifiers::PartitionId, logs::Lsn};
+
+use crate::fsm_table::{PartitionStateMachineKey, SequenceNumber, fsm_variable};
+use crate::keys::{KeyKind, TableKey};
+use crate::protobuf_types::PartitionStoreProtobufValue;
+
+const APPLIED_LSNS_PROPERTY_NAME: &CStr = c"restate_applied_lsns";
+
+#[derive(Default)]
+pub(crate) struct AppliedLsnCollector {
+    applied_lsns: HashMap<PartitionId, Lsn>,
+    properties: Vec<(CString, CString)>,
+}
+
+impl TablePropertiesCollector for AppliedLsnCollector {
+    fn add_user_key(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        entry_type: rocksdb::table_properties::EntryType,
+        _seq: u64,
+        _file_size: u64,
+    ) -> Result<(), CollectorError> {
+        if !matches!(entry_type, EntryType::EntryPut) {
+            return Ok(());
+        }
+
+        match extract_partition_applied_lsn(key, value) {
+            Ok(None) => Ok(()),
+            Ok(Some((partition_id, lsn))) => {
+                self.applied_lsns
+                    .entry(partition_id)
+                    .and_modify(|existing| {
+                        if lsn > *existing {
+                            *existing = lsn;
+                        }
+                    })
+                    .or_insert(lsn);
+                Ok(())
+            }
+            Err(err) => {
+                warn!("Failed to decode partition LSN from raw key-value: {}", err);
+                Err(CollectorError::default())
+            }
+        }
+    }
+
+    fn finish(&mut self) -> Result<impl IntoIterator<Item = &(CString, CString)>, CollectorError> {
+        if !self.applied_lsns.is_empty() {
+            let lsns_to_serialize: Vec<(PartitionId, Lsn)> =
+                self.applied_lsns.iter().map(|(k, v)| (*k, *v)).collect();
+
+            let applied_lsns_struct = AppliedLsns {
+                lsns: lsns_to_serialize,
+            };
+
+            match serde_json::to_vec(&applied_lsns_struct) {
+                Ok(json) => {
+                    self.properties.push((
+                        APPLIED_LSNS_PROPERTY_NAME.to_owned(),
+                        CString::new(json).unwrap(),
+                    ));
+                }
+                Err(err) => {
+                    warn!("Failed to serialize AppliedLsns: {}", err);
+                    return Err(CollectorError::default());
+                }
+            }
```

Comment:
I was previously using a variable key to encode multiple partitions' LSNs - a better approach is to use a fixed property name, so the read path can be dumber. This also allows us to transparently encode multiple partition/LSN pairs in the future. The only question I have is whether it's worth doing something more sophisticated about the value storage encoding here?

It would be very easy to change the encoding in the future, as we don't care about historic properties.

Reply:
As discussed, let's switch to a property-per-partition using the new `get_user_collected_property_keys` method added in https://github.com/restatedev/rust-rocksdb/pull/9/commits/2605c127121728f53958bed3af461ee54445c3e0